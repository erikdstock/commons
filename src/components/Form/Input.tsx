import React from "react"
import { Box, Card, Text, TextProps } from "rebass"
import styled from "styled-components"
import { space, themeGet } from "styled-system"

interface TextInputProps extends React.HTMLProps<HTMLInputElement> {
  borderColor?: string
  error?: string
  id: string
  noBorder?: boolean
}

const errorColor = "red"
const disabledColor = "gray.3"

export const TextInput: React.FunctionComponent<TextInputProps> = ({
  borderColor = "black",
  disabled = false,
  error = null,
  label,
  noBorder,
  ...restProps
}) => {
  let labelColor
  if (error) {
    borderColor = errorColor
    labelColor = errorColor
  } else if (disabled) {
    borderColor = disabledColor
    labelColor = disabledColor
  }
  return (
    <Wrapper m={1}>
      <Card
        py={2}
        px={2}
        borderColor={borderColor}
        border={noBorder ? "" : "2px solid"}
        borderRadius={3}
      >
        {label && (
          <InputLabel color={labelColor} htmlFor={restProps.id}>
            {label}
          </InputLabel>
        )}
        <BlankTextInput
          underline={noBorder}
          disabled={disabled}
          error={Boolean(error)}
          {...restProps as any}
        />
      </Card>
    </Wrapper>
  )
}

const Wrapper = styled(Box)`
  position: relative;
  /* border: 1px solid green; */
`

const Label: React.FunctionComponent<TextProps> = ({
  children,
  ...restProps
}) => (
  <Text
    as="label"
    bg="white"
    fontFamily="sans"
    fontSize=".75rem"
    px={1}
    {...restProps}
  >
    {children}
  </Text>
)

const InputLabel = styled(Label)`
  position: absolute;
  top: -0.4rem;
  left: ${themeGet("space.1")}px;
`

interface InputElementProps extends React.HTMLProps<HTMLInputElement> {
  underline?: boolean
  error: boolean
  disabled: boolean
}

const BlankTextInput = styled.input.attrs({
  type: "text",
})<InputElementProps>`
  border: none;
  ${({ theme, underline, error, disabled }) =>
    underline &&
    `border-bottom: solid 1px ${
      error
        ? theme.colors.red
        : disabled
        ? theme.colors[disabledColor] || "lightGray"
        : theme.colors["gray.3"] || "gray"
    }`};
  width: 100%;
  font-size: 0.85rem;
  line-height: 1.8;
  font-family: ${themeGet("fonts.sans")};
  padding-bottom: ${themeGet("space.1")};
  ${space}
  &:focus {
    outline: none;
  }
  &::placeholder {
    color: ${themeGet("colors.gray.6")};
  }
`
