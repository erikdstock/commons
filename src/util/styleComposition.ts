import styled from "@emotion/styled"
import { CSSObject, SystemStyleObject } from "@styled-system/css"
import ssCss from "@styled-system/css"
import get from "lodash/get"
import { RosesTheme } from "../theme"

/** Clumsily force styledCss function to believe that there will be a theme present */
type StyledCssFn = (o: SystemStyleObject) => CSSObject
const styledCss = ssCss as StyledCssFn

/** The base style for Box and everything on up. */
export const boxStyle: CSSObject = {
  boxSizing: "border-box",
}

interface ComposeOptions {
  defaultVariant?: string
}

const themeVariantsRoot: keyof RosesThemeObject = "variants"
const componentStylesRoot: keyof RosesThemeObject = "componentStyles"

interface StyledInterpolationProps extends RosesStyleProps {
  theme: RosesThemeObject
}

type InterpolationFn = (p: StyledInterpolationProps) => CSSObject | void

/** Reach into the theme, get a style object, apply it with styledCss */
const themedComponent: (
  path: string
) => (p: StyledInterpolationProps) => CSSObject | undefined = (
  path: string
) => (props: StyledInterpolationProps) => {
  const componentStyle = get(props.theme, `${componentStylesRoot}.${path}`)
  console.warn("CS: ", componentStyle)
  return componentStyle && styledCss(componentStyle)
}

const rxHandler: (p: StyledInterpolationProps) => CSSObject | undefined = ({
  rx,
}) => rx && styledCss(rx)

type VariantHandler = (
  key: string,
  op: ComposeOptions
) => (p: RosesStyleProps & StyledInterpolationProps) => CSSObject | undefined

const variantHandler: VariantHandler = (componentKey, ops) => ({
  theme,
  variant,
}) => {
  const selectedVariant = variant || ops.defaultVariant
  if (selectedVariant) {
    const base = `${themeVariantsRoot}.${componentKey}`
    const variantCss = get(theme, `${base}.${selectedVariant}`)
    console.warn("variant: ", variantCss)
    return variantCss && styledCss(variantCss)
  }
}

/**
 * Wrap a base styled component with:
 *   - a responsive, theme-aware style `css` prop
 *   - a `variant` prop for accessing styles set on the theme.variants key
 * @example:
 * ```tsx
 *  const RawBox = styled('div')({boxSizing: 'border-box'})
 *  const Widget = withStyleProps('Box', RawBox, {})
 * ```
 */
export const withStyleProps = (
  componentKey: string,
  Component: React.ComponentType<RosesStyleProps>,
  options: ComposeOptions = {} // not using this yet
) => {
  return styled(Component)<RosesStyleProps>(
    themedComponent(componentKey),
    variantHandler(componentKey, options),
    rxHandler
  )
}
// /** Compose in the default style + variants for the key defined in your theme
//  * @example ```
//  * const theme = {
//  *   styles: {
//  *     Button: {
//  *       color: "blue"
//  *     }
//  *   }
//  * }
//  *
//  * const Button = styled('button')(ButtonProps, themed("Button"))
//  * ```
//  */
// export const themed = key => props => props.theme.styles[key]
