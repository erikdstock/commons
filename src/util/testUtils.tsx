import React from "react"
import renderer from "react-test-renderer"
import { RosesTheme } from "../Theme"

export const render = (tree: React.ReactNode) =>
  renderer.create(<RosesTheme>{tree}</RosesTheme>)
