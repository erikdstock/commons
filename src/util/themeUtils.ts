// /** Compose in the default style + variants for the key defined in your theme
//  * @example ```
//  * const theme = {
//  *   styles: {
//  *     Button: {
//  *       color: "blue"
//  *     }
//  *   }
//  * }
//  *
//  * const Button = styled('button')(ButtonProps, themed("Button"))
//  * ```
//  */
// export const themed = key => props => props.theme.styles[key]
