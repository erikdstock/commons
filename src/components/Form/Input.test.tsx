import React from "react"
import { render } from "../../util/testUtils"
import { Input } from "./Input"

describe("Text Input", () => {
  describe("Default", () => {
    it("renders correctly", () => {
      const tree = render(
        <Input label="Name" placeholder="Hello" id="name" name="name" />
      ).toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("renders disabled state correctly", () => {
      const tree = render(
        <Input
          disabled
          label="Name"
          placeholder="Hello"
          id="name"
          name="name"
        />
      ).toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("renders error state correctly", () => {
      const tree = render(
        <Input
          error="bad praxis"
          label="Name"
          placeholder="Hello"
          id="name"
          name="name"
        />
      ).toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
  describe("No Border", () => {
    it("renders correctly", () => {
      const tree = render(
        <Input
          noBorder
          label="Name"
          placeholder="Hello"
          id="name"
          name="name"
        />
      ).toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("renders disabled state correctly", () => {
      const tree = render(
        <Input
          disabled
          noBorder
          label="Name"
          placeholder="Hello"
          id="name"
          name="name"
        />
      ).toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("renders error state correctly", () => {
      const tree = render(
        <Input
          error="bad praxis"
          noBorder
          label="Name"
          placeholder="Hello"
          id="name"
          name="name"
        />
      ).toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})
