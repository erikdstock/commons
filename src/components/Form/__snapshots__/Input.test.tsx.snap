// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Input Default renders correctly 1`] = `
.emotion-3 {
  box-sizing: border-box;
  position: relative;
  margin: 8px;
}

.emotion-2 {
  box-sizing: border-box;
  padding: 4px;
  border-radius: 3px;
  box-shadow: 0 0 16px rgba(0,0,0,.25);
  padding: 8px;
  margin: 4px;
  display: inline-block;
  border-radius: 3px;
  box-shadow: 0 0 16px rgba(0,0,0,.25);
  box-shadow: none;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 8px;
  padding-right: 8px;
  border-color: #000;
  border: 2px solid;
  border-radius: 5px;
}

.emotion-0 {
  box-sizing: border-box;
  line-height: 1.5;
  white-space: pre-wrap;
  font-family: 'Times New Roman',serif;
  background-color: #fff;
  font-family: 'Arial',system-ui,sans-serif;
  font-size: .75rem;
  padding-left: 4px;
  padding-right: 4px;
  position: absolute;
  top: -0.4rem;
  left: 4px;
}

.emotion-1 {
  border: none;
  width: 100%;
  background-color: #fff;
  font-size: 0.85rem;
  line-height: 1.8;
  font-family: 'Arial',system-ui,sans-serif;
  padding-bottom: 4px;
}

.emotion-1:focus {
  outline: none;
}

.emotion-1::-webkit-input-placeholder {
  color: hsl(0,0%,58%);
}

.emotion-1::-moz-placeholder {
  color: hsl(0,0%,58%);
}

.emotion-1:-ms-input-placeholder {
  color: hsl(0,0%,58%);
}

.emotion-1::placeholder {
  color: hsl(0,0%,58%);
}

<div
  className="emotion-3"
>
  <div
    className="emotion-2"
  >
    <label
      className="emotion-0"
      htmlFor="name"
    >
      Name
    </label>
    <input
      className="emotion-1"
      disabled={false}
      id="name"
      name="name"
      placeholder="Hello"
      type="text"
    />
  </div>
</div>
`;

exports[`Text Input Default renders disabled state correctly 1`] = `
.emotion-3 {
  box-sizing: border-box;
  position: relative;
  margin: 8px;
}

.emotion-0 {
  box-sizing: border-box;
  line-height: 1.5;
  white-space: pre-wrap;
  font-family: 'Times New Roman',serif;
  background-color: #fff;
  font-family: 'Arial',system-ui,sans-serif;
  font-size: .75rem;
  padding-left: 4px;
  padding-right: 4px;
  position: absolute;
  top: -0.4rem;
  left: 4px;
}

.emotion-1 {
  border: none;
  width: 100%;
  background-color: #fff;
  font-size: 0.85rem;
  line-height: 1.8;
  font-family: 'Arial',system-ui,sans-serif;
  padding-bottom: 4px;
}

.emotion-1:focus {
  outline: none;
}

.emotion-1::-webkit-input-placeholder {
  color: hsl(0,0%,58%);
}

.emotion-1::-moz-placeholder {
  color: hsl(0,0%,58%);
}

.emotion-1:-ms-input-placeholder {
  color: hsl(0,0%,58%);
}

.emotion-1::placeholder {
  color: hsl(0,0%,58%);
}

.emotion-2 {
  box-sizing: border-box;
  padding: 4px;
  border-radius: 3px;
  box-shadow: 0 0 16px rgba(0,0,0,.25);
  padding: 8px;
  margin: 4px;
  display: inline-block;
  border-radius: 3px;
  box-shadow: 0 0 16px rgba(0,0,0,.25);
  box-shadow: none;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 8px;
  padding-right: 8px;
  border-color: hsl(0,0%,58%);
  border: 2px solid;
  border-radius: 5px;
}

<div
  className="emotion-3"
>
  <div
    className="emotion-2"
  >
    <label
      className="emotion-0"
      color="gray.3"
      htmlFor="name"
    >
      Name
    </label>
    <input
      className="emotion-1"
      disabled={true}
      id="name"
      name="name"
      placeholder="Hello"
      type="text"
    />
  </div>
</div>
`;

exports[`Text Input Default renders error state correctly 1`] = `
.emotion-3 {
  box-sizing: border-box;
  position: relative;
  margin: 8px;
}

.emotion-0 {
  box-sizing: border-box;
  line-height: 1.5;
  white-space: pre-wrap;
  font-family: 'Times New Roman',serif;
  background-color: #fff;
  font-family: 'Arial',system-ui,sans-serif;
  font-size: .75rem;
  padding-left: 4px;
  padding-right: 4px;
  position: absolute;
  top: -0.4rem;
  left: 4px;
}

.emotion-1 {
  border: none;
  width: 100%;
  background-color: #fff;
  font-size: 0.85rem;
  line-height: 1.8;
  font-family: 'Arial',system-ui,sans-serif;
  padding-bottom: 4px;
}

.emotion-1:focus {
  outline: none;
}

.emotion-1::-webkit-input-placeholder {
  color: hsl(0,0%,58%);
}

.emotion-1::-moz-placeholder {
  color: hsl(0,0%,58%);
}

.emotion-1:-ms-input-placeholder {
  color: hsl(0,0%,58%);
}

.emotion-1::placeholder {
  color: hsl(0,0%,58%);
}

.emotion-2 {
  box-sizing: border-box;
  padding: 4px;
  border-radius: 3px;
  box-shadow: 0 0 16px rgba(0,0,0,.25);
  padding: 8px;
  margin: 4px;
  display: inline-block;
  border-radius: 3px;
  box-shadow: 0 0 16px rgba(0,0,0,.25);
  box-shadow: none;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 8px;
  padding-right: 8px;
  border-color: #db1d33;
  border: 2px solid;
  border-radius: 5px;
}

<div
  className="emotion-3"
>
  <div
    className="emotion-2"
  >
    <label
      className="emotion-0"
      color="red"
      htmlFor="name"
    >
      Name
    </label>
    <input
      className="emotion-1"
      disabled={false}
      id="name"
      name="name"
      placeholder="Hello"
      type="text"
    />
  </div>
</div>
`;

exports[`Text Input No Border renders correctly 1`] = `
.emotion-3 {
  box-sizing: border-box;
  position: relative;
  margin: 8px;
}

.emotion-0 {
  box-sizing: border-box;
  line-height: 1.5;
  white-space: pre-wrap;
  font-family: 'Times New Roman',serif;
  background-color: #fff;
  font-family: 'Arial',system-ui,sans-serif;
  font-size: .75rem;
  padding-left: 4px;
  padding-right: 4px;
  position: absolute;
  top: -0.4rem;
  left: 4px;
}

.emotion-2 {
  box-sizing: border-box;
  padding: 4px;
  border-radius: 3px;
  box-shadow: 0 0 16px rgba(0,0,0,.25);
  padding: 8px;
  margin: 4px;
  display: inline-block;
  border-radius: 3px;
  box-shadow: 0 0 16px rgba(0,0,0,.25);
  box-shadow: none;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 8px;
  padding-right: 8px;
  border-color: #000;
  border: none;
  border-radius: 5px;
}

.emotion-1 {
  border: none;
  width: 100%;
  background-color: #fff;
  font-size: 0.85rem;
  line-height: 1.8;
  font-family: 'Arial',system-ui,sans-serif;
  padding-bottom: 4px;
  border-bottom: solid 1px;
}

.emotion-1:focus {
  outline: none;
}

.emotion-1::-webkit-input-placeholder {
  color: hsl(0,0%,58%);
}

.emotion-1::-moz-placeholder {
  color: hsl(0,0%,58%);
}

.emotion-1:-ms-input-placeholder {
  color: hsl(0,0%,58%);
}

.emotion-1::placeholder {
  color: hsl(0,0%,58%);
}

<div
  className="emotion-3"
>
  <div
    className="emotion-2"
  >
    <label
      className="emotion-0"
      htmlFor="name"
    >
      Name
    </label>
    <input
      className="emotion-1"
      disabled={false}
      id="name"
      name="name"
      placeholder="Hello"
      type="text"
    />
  </div>
</div>
`;

exports[`Text Input No Border renders disabled state correctly 1`] = `
.emotion-3 {
  box-sizing: border-box;
  position: relative;
  margin: 8px;
}

.emotion-0 {
  box-sizing: border-box;
  line-height: 1.5;
  white-space: pre-wrap;
  font-family: 'Times New Roman',serif;
  background-color: #fff;
  font-family: 'Arial',system-ui,sans-serif;
  font-size: .75rem;
  padding-left: 4px;
  padding-right: 4px;
  position: absolute;
  top: -0.4rem;
  left: 4px;
}

.emotion-1 {
  border: none;
  width: 100%;
  background-color: #fff;
  font-size: 0.85rem;
  line-height: 1.8;
  font-family: 'Arial',system-ui,sans-serif;
  padding-bottom: 4px;
  border-bottom: solid 1px;
}

.emotion-1:focus {
  outline: none;
}

.emotion-1::-webkit-input-placeholder {
  color: hsl(0,0%,58%);
}

.emotion-1::-moz-placeholder {
  color: hsl(0,0%,58%);
}

.emotion-1:-ms-input-placeholder {
  color: hsl(0,0%,58%);
}

.emotion-1::placeholder {
  color: hsl(0,0%,58%);
}

.emotion-2 {
  box-sizing: border-box;
  padding: 4px;
  border-radius: 3px;
  box-shadow: 0 0 16px rgba(0,0,0,.25);
  padding: 8px;
  margin: 4px;
  display: inline-block;
  border-radius: 3px;
  box-shadow: 0 0 16px rgba(0,0,0,.25);
  box-shadow: none;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 8px;
  padding-right: 8px;
  border-color: hsl(0,0%,58%);
  border: none;
  border-radius: 5px;
}

<div
  className="emotion-3"
>
  <div
    className="emotion-2"
  >
    <label
      className="emotion-0"
      color="gray.3"
      htmlFor="name"
    >
      Name
    </label>
    <input
      className="emotion-1"
      disabled={true}
      id="name"
      name="name"
      placeholder="Hello"
      type="text"
    />
  </div>
</div>
`;

exports[`Text Input No Border renders error state correctly 1`] = `
.emotion-3 {
  box-sizing: border-box;
  position: relative;
  margin: 8px;
}

.emotion-0 {
  box-sizing: border-box;
  line-height: 1.5;
  white-space: pre-wrap;
  font-family: 'Times New Roman',serif;
  background-color: #fff;
  font-family: 'Arial',system-ui,sans-serif;
  font-size: .75rem;
  padding-left: 4px;
  padding-right: 4px;
  position: absolute;
  top: -0.4rem;
  left: 4px;
}

.emotion-2 {
  box-sizing: border-box;
  padding: 4px;
  border-radius: 3px;
  box-shadow: 0 0 16px rgba(0,0,0,.25);
  padding: 8px;
  margin: 4px;
  display: inline-block;
  border-radius: 3px;
  box-shadow: 0 0 16px rgba(0,0,0,.25);
  box-shadow: none;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 8px;
  padding-right: 8px;
  border-color: #db1d33;
  border: none;
  border-radius: 5px;
}

.emotion-1 {
  border: none;
  width: 100%;
  background-color: #fff;
  font-size: 0.85rem;
  line-height: 1.8;
  font-family: 'Arial',system-ui,sans-serif;
  padding-bottom: 4px;
  border-bottom: solid 1px;
  border-color: #db1d33;
}

.emotion-1:focus {
  outline: none;
}

.emotion-1::-webkit-input-placeholder {
  color: hsl(0,0%,58%);
}

.emotion-1::-moz-placeholder {
  color: hsl(0,0%,58%);
}

.emotion-1:-ms-input-placeholder {
  color: hsl(0,0%,58%);
}

.emotion-1::placeholder {
  color: hsl(0,0%,58%);
}

<div
  className="emotion-3"
>
  <div
    className="emotion-2"
  >
    <label
      className="emotion-0"
      color="red"
      htmlFor="name"
    >
      Name
    </label>
    <input
      className="emotion-1"
      disabled={false}
      id="name"
      name="name"
      placeholder="Hello"
      type="text"
    />
  </div>
</div>
`;
